paimon-core/src/main/java/org/apache/paimon/manifest/ManifestFileMeta.java

```diff
@@ -185,4 +209,138 @@ private static void mergeCandidates(
+        int i;
+        for (i = 0; i < inputs.size(); i++) {
+            ManifestFileMeta manifestFileMeta = inputs.get(i);
+            if (manifestFileMeta.numDeletedFiles == 0
+                    && manifestFileMeta.fileSize >= suggestedMetaSize) {
+                base.add(manifestFileMeta);
+                totalManifestSize += manifestFileMeta.fileSize;
+            } else {
+                break;
+            }
+        }
+
+        List<ManifestFileMeta> delta = new ArrayList<>();
+        for (; i < inputs.size(); i++) {
+            ManifestFileMeta manifestFileMeta = inputs.get(i);
+            delta.add(manifestFileMeta);
+            totalManifestSize += manifestFileMeta.fileSize;
+        }
+
+        long deltaDeleteFileNum = 0;
+        long totalDeltaFileSize = 0;
+        for (ManifestFileMeta manifestFileMeta : delta) {
+            deltaDeleteFileNum += manifestFileMeta.numDeletedFiles();
+            totalDeltaFileSize += manifestFileMeta.fileSize();
+            totalManifestSize += manifestFileMeta.fileSize();
+        }
+
+        if (totalDeltaFileSize < sizeTrigger) {
+            return inputs;
+        }
+
+        logger.info(
+                "start Manifest File Full Compaction,pick the number of delete file: {} ,total manifest file size: {}",
+                deltaDeleteFileNum,
+                totalManifestSize);
+        // do full-compaction
+        Map<ManifestEntry.Identifier, ManifestEntry> deltaMerged = new LinkedHashMap<>();
+        for (ManifestFileMeta manifest : delta) {
+            ManifestEntry.mergeEntries(manifestFile.read(manifest.fileName), deltaMerged);
+        }
+        Set<BinaryRow> deletePartitions = computeDeletePartitions(deltaMerged);
+
+        Predicate predicate = convertPartitionToPredicate(partitionType, deletePartitions);
+
+        List<ManifestFileMeta> result = new ArrayList<>();
+        FieldStatsArraySerializer fieldStatsArraySerializer =
+                new FieldStatsArraySerializer(partitionType);
+        int j;
+        for (j = 0; j < base.size(); j++) {
+            // todo  optimize this to binary search.
+            ManifestFileMeta file = base.get(j);
+            if (predicate.test(
+                    file.numAddedFiles + file.numDeletedFiles,
+                    file.partitionStats.fields(fieldStatsArraySerializer))) {
+                break;
+            } else {
+                result.add(file);
+            }
+        }
+
+        Map<ManifestEntry.Identifier, ManifestEntry> fullMerged = new LinkedHashMap<>();
+        for (; j < base.size(); j++) {
+            ManifestFileMeta manifestFileMeta = base.get(j);
+            ManifestEntry.mergeEntries(manifestFile.read(manifestFileMeta.fileName), fullMerged);
+        }
+
+        ManifestEntry.mergeEntries(deltaMerged.values(), fullMerged);
+        if (!fullMerged.isEmpty()) {
+            List<ManifestFileMeta> merged =
+                    manifestFile.write(new ArrayList<>(fullMerged.values()));
```

Comment:
maybe, before write full merged manifest entries, we can sort them by partition. This allows to obtain better efficiency of data skipping.

Reply:
Or we can just create an issue for this.
I'm not sure if there are side effects, maybe we need to add an OPTION to explore.

Reply:
Maybe we can reuse this option `manifest.full-compaction-threshold-size`. Set the default value of the option to Long.MaxValue

Reply:
> This will result in the order of ManifestEntry being affected. Perhaps we need some logic to merge ManifestEntry. I'm not entirely sure if this will have an impact on queries,